### PR TITLE
Report model predict time in megapixel bucket durations

### PR DESCRIPTION
**Pull request by SolomonLake**
## What does this PR do?

Linked issue: [LAKE-811](https://linear.app/roboflow/issue/LAKE-811/megapixel-buckets-use-model-infer-time-rather-than-also-processing)

The `megapixel_buckets` `execution_duration` on model usage rows carried the usage decorator's full wall clock around `infer()`. That window includes image decoding (a URL fetch lands inside it), resize and normalization, NMS, mask-to-polygon conversion and response construction — work that is largely independent of model size. Because that overhead is roughly constant, it compresses the measured gap between small and large models and, combined with differences in deployment hardware and input resolution, can make a nano model look slower than a large one.

Bucket durations now report the model's predict phase alone. `BaseInference.infer` times the existing `model.predict` span and publishes the result through a `ContextVar`, reusing the seam that already carries preprocess input size out to the usage decorator — ContextVars rather than model attributes because model instances are shared across worker threads. `get_model_megapixel_buckets` consumes that value and falls back to the full call duration when no predict phase was published, which covers the families that override `infer()` wholesale and the SAM `infer_from_request` entrypoints. Calls whose `predict()` returns a pending result also fall back, so the pipelined RF-DETR segmentation adapter never reports a kernel launch as if it were inference time. The row-level `execution_duration` is unchanged and still reports wall clock, so bucket durations no longer reconcile against the row total — the difference between them is now the pre/post overhead. Since this replaces a value rather than adding a key, `merge_megapixel_buckets` and its redis-offloader mirror need no change.

**Main elements:**
- `inference/usage_tracking/megapixel_buckets.py` — new `_measured_predict_duration` ContextVar with `record_measured_predict_duration` / `consume_measured_predict_duration`, the `prediction_is_deferred` guard, and `clear_measured_model_input` extended to reset both measurements
- `inference/core/models/base.py` — times the predict phase inside the existing `model.predict` span
- `inference/usage_tracking/decorator_helpers.py` — `get_model_megapixel_buckets` prefers the published predict duration, consuming it before the test-run check so an unreported call cannot leak into the next one
- No change to usage payload shape, `merge_megapixel_buckets`, or the row-level `execution_duration`

## Type of Change

- [ ] Bug fix (non-breaking change that fixes an issue)
- [ ] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update
- [ ] Refactoring (no functional changes)
- [x] Other: telemetry semantics change (same payload shape, different meaning for bucket `execution_duration`)

### Unit tests

- `record_measured_predict_duration` accumulates across multiple predict calls in one decorated call, and drops negative / non-numeric values
- `consume_measured_predict_duration` clears on read, so a call with no predict phase cannot inherit an earlier measurement
- `clear_measured_model_input` resets the predict duration alongside the input size
- Bucket duration prefers a published predict duration and falls back to the call duration without one
- The published duration is consumed even for an unbucketed test run
- End-to-end through `BaseInference.infer`: a model sleeping in preprocess and postprocess reports a bucket duration matching only its predict phase
- `prediction_is_deferred` detects future-like results and not tuples, arrays or dicts
- A model whose `predict()` returns a pending result publishes nothing and falls back
- Through the usage decorator: the row keeps wall clock while the bucket narrows to predict, and a model with no predict phase has bucket duration equal to the row's

### Integration tests

- None for this PR.

### Other

- `pytest tests/inference/unit_tests/usage_tracking/ tests/inference/unit_tests/core/models/` — 236 passed
- `black --check` and `isort --check-only` on all changed files

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code where necessary, particularly in hard-to-understand areas
- [x] My changes generate no new warnings or errors
- [ ] I have updated the documentation accordingly (if applicable)

## Additional Context

Two known limits, both documented in the `megapixel_buckets` module docstring:

- **Async CUDA launches.** The predict phase is timed on the host, so a backend that launches kernels without synchronizing before `predict()` returns still under-reports; the wait lands in whichever later phase forces synchronization. Deferred-result backends are excluded explicitly, but plain async launches cannot be detected here. An opt-in `torch.cuda.synchronize()` at the phase boundary would close this if it proves material.
- **Coverage.** Only `BaseInference.infer` is instrumented. DocTR, GroundingDINO, YOLOWorld, OwlV2 and RFInstant override `infer()` without a phase split, and SAM2/SAM3 decorate `infer_from_request`; all fall back to wall clock, so comparisons spanning those families are not yet like-for-like. `test_model_decorator_buckets_full_duration_without_predict_phase` pins the fallback so it stays deliberate.

Separately, for `RFDETR_PIPELINE_DEPTH >= 2` the row's wall-clock duration was already only loosely coupled to the frame it is attributed to, since a pipelined call can return an earlier frame's responses or empty ones. That predates this change, but those deployments should be excluded from latency comparisons.

Made with [Cursor](https://cursor.com)

### CODE REVIEW

File: inference/core/models/base.py [hunk 1]
```diff
@@ -42,7 +44,10 @@ def infer(self, image: Any, **kwargs) -> Any:
             if hasattr(preproc_image, "shape"):
                 set_span_attribute("model.input_shape", str(preproc_image.shape))
         with start_span("model.predict"):
+            predict_started_at = perf_counter()
             predicted_arrays = self.predict(preproc_image, **kwargs)
+            if not prediction_is_deferred(predicted_arrays):
+                record_measured_predict_duration(perf_counter() - predict_started_at)
```
[the remark this answers is not in the document]
Comment (SolomonLake) [reply inference/core/models/base.py]: The trace is right, and following it turned up a second bug I would otherwise have shipped. Addressed in d2be9fdc and f4fe69f7 (not auto-reviewed).

**What changed.** Rather than synchronizing at the phase boundary (option 1, which puts a device sync on the request path) or restricting publication to known-synchronous entrypoints (option 2, which leaves the primary GPU path on wall clock), the phase is now bracketed with CUDA events on the stream the model runs on, and read **after** post-processing rather than after `predict()`. Reading two events requires the second to have completed — and by the time postprocess has done the `det.xyxy.detach().cpu().numpy()` you identified at `inference_models_adapters.py:298-300`, it has. So the read finds both events complete and never waits. If an event is somehow incomplete the reading is dropped in favour of the host clock rather than blocked on, so no path here can add latency.

**The second bug.** Chasing your trace into the TRT backend: `rfdetr_object_detection_trt.py:675` hands engine execution to `self._scheduler.execute_engine(...)`, and `BaseExecutionScheduler` owns `self._inference_stream = torch.cuda.Stream(device=device)` (`optimization/schedulers/base.py:67`) — a dedicated stream, not the current one. My first pass only looked for `_inference_stream` on the model and the backend it wraps, so RF-DETR TRT fell through to guessing `torch.cuda.current_stream(device)`, where nothing had been enqueued. That reads near zero: the same inversion you describe, produced by a third mechanism. Resolution now searches a scheduler at each level, and guessing is bounded — a *declared* stream is known to carry the model's work and replaces the host window outright, while an *assumed* one is believed only when it exceeds that window, where it can only be revealing device work that outlived `predict()`. A wrong guess can therefore be ignored but never believed.

**Coverage now.** Declared streams (the ~45 ONNX/TRT backends that expose `_inference_stream`, plus RF-DETR TRT via its scheduler) get device timing. The RF-DETR PyTorch backend manages no stream, so it is the assumed case and gets device timing only when the reading proves the host window was too short. Legacy ONNX stays on the host clock, which is already correct there because `run_session_via_iobinding` binds outputs to host buffers and returns numpy.

**Caveat I would rather state than paper over:** none of this is exercised against a real GPU — the tests drive fake events, and there is no GPU in the loop here. The fallbacks are the protection: every uncertain path lands on the host clock, which is the number this metric reported before the change.

On the session lock: agreed, and it is real for legacy ONNX. I have recorded it in the module docstring rather than changed behaviour, since it predates this PR:

> What is measured is the phase, not the model in isolation. Legacy ONNX models take their session lock inside `predict()`, so a host reading there carries time spent queued behind other requests to the same model. That was true of the wall-clock duration this replaces as well, but it is contention rather than model cost sitting inside a number now labelled as the model's.

File: inference/usage_tracking/decorator_helpers.py [hunk 2]
```diff
@@ -154,16 +155,43 @@ def get_model_megapixel_buckets(
...
+    """
+    predict_duration = consume_measured_predict_duration()
     if inference_test_run:
         return {}
+
     height, width = input_hw if input_hw else (None, None)
-    return build_megapixel_buckets(
+    bucket_duration = (
+        predict_duration if predict_duration is not None else execution_duration
+    )
```
[the remark this answers is not in the document]
Comment (SolomonLake) [reply inference/usage_tracking/decorator_helpers.py:185]: Owner decision: not adding a discriminator, because these numbers are analytics-only and no billable or quota path reads bucket durations. Flagging what remains true so it is on the record rather than dismissed.

One correction to the framing, though, because it narrows the finding: the fallback is usually **not** full wall clock. Where the timer publishes a host reading, that reading is the clock around `predict()` alone, which for every backend that synchronizes before returning *is* the predict phase. The genuine full-call-duration cases are only the two where nothing is published at all: deferred results, and the families that never reach this code path (OWLv2 / DocTR / GroundingDINO / YOLOWorld / RFInstant overriding `infer()`, SAM2/SAM3 decorating `infer_from_request`).

After d2be9fdc and f4fe69f7 the mixed-semantics surface is smaller still, since the `inference_models` GPU backends now measure the device rather than the launch. But your sharpest point survives all of that: the same `resource_id` flips between predict-only and full-call depending on `RFDETR_PIPELINE_DEPTH` and on batch size taking the non-pipelined branch, and a consumer cannot see which served the call. That is a real limitation of the current payload, accepted deliberately here rather than solved.

If the cross-model comparison this feeds turns out to need it, the additive shape you suggest is the right one — a separate key alongside a wall-clock `execution_duration`, which keeps the merge helpers and the redis mirror pure additions — and it should be a change made together with the platform side rather than landed blind from this repo.

**Comment by SolomonLake** [comment]
Answers to both, from the owner side.

**1. Does the downstream pipeline distinguish predict-only durations from wall-clock fallbacks?** No, and it does not need to — dropping this one. Since the question was written the fallback surface has also narrowed: d2be9fdc and f4fe69f7 moved the `inference_models` GPU backends onto CUDA-event timing, so they now measure the device rather than the kernel launch, which was the largest source of rows whose meaning did not match the label. Detail in the reply on `inference/core/models/base.py`. What remains is genuinely mixed — a pipelined RF-DETR flips on `RFDETR_PIPELINE_DEPTH` and batch size, and the override-`infer()` families never publish — and that is accepted rather than solved; see the reply on `decorator_helpers.py`.

**2. Is anything deriving billing, quota, or SLO numbers from bucket `execution_duration`?** No. Bucket durations are analytics-only, and every billable and quota path reads the row-level `execution_duration`, which is untouched by this PR and still carries the `GCP_SERVERLESS` floor from `_compute_execution_duration`. A sub-10ms predict phase landing in the bucket as sub-10ms is the intended behaviour — the floor exists to protect billing, and applying it to an analytics measurement would defeat the point of measuring, since it would quantize exactly the fast models this metric is meant to compare.

Both of the tests suggested in the review summary are now in `tests/inference/unit_tests/usage_tracking/test_predict_timing.py`: `test_predict_duration_does_not_leak_when_postprocessing_raises` and `test_predict_durations_do_not_leak_between_threads`.

On the minor doubts: the flaky assertion is now relative to the measured call duration rather than an absolute `< 0.05`, so a loaded runner that inflates every sleep no longer fails it. The order-sensitivity one is already covered — `test_megapixel_buckets.py` has an autouse fixture calling `clear_measured_model_input()`, which clears the predict duration too. The session-lock observation is recorded in the module docstring, quoted in the `base.py` reply.

File: inference/usage_tracking/predict_timing.py [hunk 3]
```diff
@@ -0,0 +1,316 @@
...
+        try:
+            start_event, end_event = self._events
+            if not start_event.query() or not end_event.query():
+                return None
+            seconds = start_event.elapsed_time(end_event) / 1000.0
+            if seconds < 0 or seconds > perf_counter() - self._started_at:
+                return None
+            if not self._stream_is_declared and seconds <= self._host_duration:
+                return None
+            return seconds
```
[the remark this answers is not in the document]
Comment (SolomonLake) [reply inference/usage_tracking/predict_timing.py:264]: Taking the bound, not just the caveat.

Confirmed against `infer_from_trt_engine`: it defaults to `synchronize=True` and calls `stream.synchronize()` before returning. The `synchronize=False` path is the pipelined one, which already publishes nothing via `prediction_is_deferred`. So a declared reading past this call's host window is the shared-stream interleaving you traced, not async compute the host missed.

A declared reading that exceeds `_host_duration` is now dropped (same published value as `min(seconds, self._host_duration)` — `publish()` already falls back to the host clock). Uncontended calls still narrow when the device reading is shorter. The module docstring now covers per-instance stream contention the same way it already covered the legacy ONNX session lock.

Pinned by `test_declared_stream_reading_is_ignored_when_it_exceeds_the_host_window` (the gap the full-call upper bound misses: inflated reading that still finishes during post-processing) and `test_shared_declared_stream_does_not_absorb_another_calls_window`. The existing undercut test stays, renamed to `test_declared_stream_reading_is_used_when_it_undercuts_the_host_window`, so both sides of the declared-stream bound are explicit.

**Comment by SolomonLake** [comment]
not sure if we need these changes, closing for now
